### ISSUE TITLE
Return array value

### ISSUE THREAD

**Issue by njtrettel** [ballercat/walt#125]
# Feature Request
## Overview
Not sure if this is a bug or not, I read through the docs and found no reference to this. Also tried looking through issues.

I can't return an array. Correct me if i'm off here, but I think it's because behind the scenes, walt is just grabbing the memory reference for the first element of the array. Similar to C if you print `str` (where str is an `char[]`), you'd just get the first char.

This is different from the issue about strings.

# Impact
Large (if it doesn't already exist).

## Details
Here's a dumb little example I tried:
```js
const memory: Memory<{ initial: 1}>;

export function sortItems(array: i32[]): i32[] {
  let i: i32;
  let toReturn: i32[] = 0;
  for (i = 0; i < 5; i += 1) {
    toReturn[i] = i;
  }
  return toReturn;
}
```
I get `Inconsistent return value. Expected i32[], received i32`
## Due Date
¯\\_(ツ)_/¯
7/20/18
if I missed something, I guess there is no due date.

I'm working on this for a hackday (which is only today), so either way, I'll unfortunately have to ditch this project. I'm going to just use the C -> WASM compiler. However, I opened this issue because I'd like to experiment more in the future.

**Comment by ballercat** [comment]
Hey, thanks for opening an issue!

Arrays are more like raw C pointers to the start address. So you can just change the return value of `sortItems` to be `i32`. That is also why `toReturn: i32[] = 0;` makes sense(though is dangerous, because you are overriding anything at address 0)

The `i32[]` is more of a compiler hint than anything.

**Comment by njtrettel** [comment]
Thanks for the response!

I can grasp how that would work in C - it's exactly like it should be done in C. But I don't know how I would be able to access the rest of that array in JS. 

I changed it to return simply i32. When I call my WASM module in JS, the return value is just an integer (as the return value suggests). In C, I would just reference `myReturnValue[i * sizeof(int)]`. But we don't have that in JS.

**Comment by ballercat** [comment]
I should also mention that to access the memory of a wasm module from JS, it first needs to be provided _by the JS bootstrap logic_. Similar to an import you had in your other issue.

**Comment by njtrettel** [comment]
@ballercat you've totally saved my entire day, thank you for your help on both of these issues I opened. I'm in the middle of attempting your solution above ^^, but from first glance, it looks like it'll be fine.

**Comment by njtrettel** [comment]
One last question. You said it needs to be passed in along with the bootstrapper. Looking at the `instantiate` docs, it doesn't seem obvious where I would pass that in. Can you explain that a little further?

**Comment by ballercat** [comment]
Sure, all I mean is that it needs to be part of the imports to the WASM module. Per wasm spec, a Memory reference in a WebAssembly module is either declared or _imported_ as a regular import with a memory type.


```js
const memory = new WebAssembly.Memory({ initial: 1 });
makeSorter({
  env: {
    memory,
    log: console.log
  }
}).then(module => ...);
```

walt source, but really anything that compiles to wasm
```
import { memory: Memory<{ initial: 1 }> } from 'env';
```

**Comment by njtrettel** [comment]
Okay, sorry for the long thread of questions. Now I've got the memory object imported from `env` in the walt code, and I have created a DataView with the buffer to read from. But whenever I try to do array operations in walt, I get an out of bounds error. I think I'm just missing something again.

Apologies if this is documented. I searched for it and couldn't find it here. The WebAssembly docs don't help much in this case.
